### ISSUE TITLE
Google-style: opening and closing quotes of text-blocks should be on new lines

### ISSUE THREAD

**Issue by mohitsatr** [checkstyle/checkstyle#17159]
I have read check documentation: https://checkstyle.org/checks/xxxxxx/nameofaffectedcheck.html
I have downloaded the latest cli from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

> https://google.github.io/styleguide/javaguide.html#s4.8.9-text-blocks
The opening """ of a text block is always on a new line. That line may either follow the same indentation rules as other constructs, or it may have no indentation at all (so it starts at the left margin). The closing """ is on a new line with the same indentation as the opening """, and may be followed on the same line by further code

```
/** some javadoc.  */
public class TextBlocks {

  /** some javadoc. */
  public static String textFun(String[] args) {
                                
    final String simpleScript = """ // EXPECTED VIOLATION
        s
        """;

    final String simpleScript1 = 
        """
        this is simple test;
        """;

    final String simpleScript2 = 
        """
        this is sample text """;   // EXPECTED VIOLATION
        
    return """                     // EXPECTED VIOLATION
        this is sample text 
        """;
  }
}


$ java -jar testing/checkstyle-10.25.0-all.jar -c testing/google_checks.xml TextBlocks.java 
Starting audit...
Audit done.
```
We currently do not fulfill this requirement from the style-guide.

**Comment by mohitsatr** [comment]
@Zopsss we need take this nuance into consideration too for https://github.com/checkstyle/checkstyle/issues/11867 issue

**Comment by Zopsss** [comment]
Good catch! We can use [MatchXpath](https://checkstyle.org/checks/coding/matchxpath.html#MatchXpath) to implement this rule.

**Comment by romani** [comment]
Looks like very formatting issues, please show formatter behavior on this code.
If formatter doesn't enforce it, let's recheck their issue tracker, we can open issue on formatter.

**Comment by mohitsatr** [comment]
@romani I updated the description. Formatter behaves correctly except when opening quotes are preceded by `return`.

**Comment by Zopsss** [comment]
```java
return """   //INCORRECT, should be on new new line
    this is sample text
    """;
``` 

Please report this to google-java-format

**Comment by mohitsatr** [comment]
https://github.com/google/google-java-format/issues/1258

**Comment by Zopsss** [comment]
Good! In your free time, please try to solve it using `MatchXpath`

**Comment by romani** [comment]
why xpath ?
We can make Check that demand `"""` to be on new line.

please add all of this to our Inputs before  fix of issue.
please add:
```
//OK , """ should be on new new line
return 
    """
    this is sample text
    """;
```

please add to issue example of checkstyle execution with google style.

**Comment by Zopsss** [comment]
> We can make Check that demand """ to be on new line.

Do we really need a new check for this? Is this in demand by other users? Or other style guides also require this?

**Comment by romani** [comment]
we can reuse MatchXpath, but we should not limit ourself to use it, if implementation become challenging. I do not think we have easy have by xpath expressions validate that some token is only token on a line.

**Comment by Zopsss** [comment]
> I do not think we have easy have by xpath expressions validate that some token is only token on a line.

I think we should first try with xpath approach, if that doesn't work then we can create a separate check for it. Checks requires lot of time, so if we're able to find a quicker solution then I think we should go for it instead of a new check.

**Comment by mohitsatr** [comment]
@Zopsss how Xpath ? we want to validate if the quotations are preceded and followed by something. 
for example 
```
String p1 = """  // Preceded by some code, not valid
content...
""";

String p2 = 
""" content of text-block abc ..          //Followed by content, Not valid
""";

String p3 =
"""    // OK
some content
""";
```
shouldn't using regexp makes more sense in this case?
I will try and see what happens .

**Comment by Zopsss** [comment]
By Xpath I meant MatchXpath Check: https://checkstyle.org/checks/coding/matchxpath.html#MatchXpath

It allows us to create custom checks using xpath queries, in our case we want to give violation for `"""` if it doesn't match google style, currently we do not have such Check, so what I was suggesting was to use MatchXpath Check to implement this rule. If in future, more demand comes for multi-line string then we can think of creating a separate check for it so everybody can easily use it but for now MatchXpath should be sufficient.

**Comment by mohitsatr** [comment]
In xpath query we can only access a node and it's children, but cannot determine their line no. 
so let's say we want to check if `TEXT_BLOCK_LITERAL_BEGIN` is the only thing on the new line, we will compare it's line no. with `VARIABLE_DEF` . and that's not possible.
am I missing something?

**Comment by mohitsatr** [comment]
The following pattern works, but it gives violation when the pattern matches the code, and we want the opposite case i.e give violation when a line does not match the pattern

```    
<module name="RegexpMultiline">
        <property name="id" value="TripleQuotePrecededByCode"/>
        <property name="format" value='^\s*"""$'/>
        <property name="message" value="Opening triple-quote must be on its own line, not preceded by code."/>
    </module>
```
is there a way to do this ?

**Comment by Zopsss** [comment]
> am I missing something?

my bad, it might be possible, idk, but if it is possible then it would be very difficult to write such query

**Comment by Zopsss** [comment]
> is there a way to do this ?

Try this:

```xml
<module name="RegexpMultiline">
        <property name="id" value="TripleQuotePrecededByCode"/>
        <property name="format" value='[\S].*&quot;&quot;&quot;'/>
        <property name="message" value="triple quotes must not be followed by characters."/>
    </module>
```

```java
public class Test {
  // valid cases
  String s1 =
      """
       Testing....
      """;
  String s2 =
      """
       Testing....
      """ + "closing quotes can be followed by code...";

  // invalid cases
  String s3 = """
       Testing....
      """;
  String s4 =
      """ 
      Testing....
      testing..  """;
}
```

```
ava -jar .\checkstyle-10.25.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] C:\checkstyle testing\Test.java:13: triple quotes must be alone on the line. [TripleQuotePrecededByCode]
[ERROR] C:\checkstyle testing\Test.java:19: triple quotes must be alone on the line. [TripleQuotePrecededByCode]
Audit done.
Checkstyle ends with 2 errors.
```

`s3`'s opening `"""` gives violation and `s4`'s closing `"""` gives violation, both are expected.

I roughly made this regex and tested it, if you find any edge cases then please let us know. 

Also try to improve the error message. @mohitsatr

**Comment by romani** [comment]
>The closing """ is on a new line with the same indentation as the opening """

We need special Check, Regexp is not an option

**Comment by Zopsss** [comment]
@mohitsatr please open an issue about the new check

**Comment by Zopsss** [comment]
> We need special Check, Regexp is not an option

Until new check is implemented, should we use Regexp to partially match the rule? We can add the blue tick in coverage table and reference the New Check's issue with a description like: 

> The rule is partially implemented, currently we're not able to check indentation of closing `"""`. The rule will be fully supported once the [CheckName]Check is implemented: #IssueNumber 

The message wording can be improved. @romani

**Comment by romani** [comment]
>Until new check is implemented, should we use Regexp to partially match the rule?

No need. Nobody reported this gap in coverage, so it can wait for sure

**Comment by DakshRJain737** [comment]
The new check is already implemented, Is this issue still open for work??

**Comment by romani** [comment]
https://checkstyle.org/google_style.html#a4.8.9, hmm, yes, and it is even ativated in google style config.


```
/var/tmp$ cat -n Test.java 
     1	/** some javadoc.  */
     2	public class TextBlocks {
     3	
     4	  /** some javadoc. */
     5	  public static String textFun(String[] args) {
     6	                                
     7	    final String simpleScript = """ // EXPECTED VIOLATION
     8	        s
     9	        """;
    10	
    11	    final String simpleScript1 = 
    12	        """
    13	        this is simple test;
    14	        """;
    15	
    16	    final String simpleScript2 = 
    17	        """
    18	        this is sample text """;   // EXPECTED VIOLATION
    19	        
    20	    return """                     // EXPECTED VIOLATION
    21	        this is sample text 
    22	        """ + simpleScript1;  // OK if followed  by some code.
    23	  }
    24	}

/var/tmp$ java -jar checkstyle-13.0.0-all.jar -c /google_checks.xml Test.java 
Starting audit...
[WARN] /var/tmp/Test.java:1: Forbidden summary fragment. [SummaryJavadoc]
[WARN] /var/tmp/Test.java:2:1: The name of the outer type and the file do not match. [OuterTypeFilename]
[WARN] /var/tmp/Test.java:4: Forbidden summary fragment. [SummaryJavadoc]
[WARN] /var/tmp/Test.java:7:33: Opening quotes (""") of text-block must be on the new line. [TextBlockGoogleStyleFormatting]
[WARN] /var/tmp/Test.java:9:9: Text-block quotes are not vertically aligned. [TextBlockGoogleStyleFormatting]
[WARN] /var/tmp/Test.java:18:29: Closing quotes (""") of text-block should not be preceded by non-whitespace characters. [TextBlockGoogleStyleFormatting]
[WARN] /var/tmp/Test.java:18:29: Text-block quotes are not vertically aligned. [TextBlockGoogleStyleFormatting]
[WARN] /var/tmp/Test.java:20:12: Opening quotes (""") of text-block must be on the new line. [TextBlockGoogleStyleFormatting]
[WARN] /var/tmp/Test.java:22:9: Text-block quotes are not vertically aligned. [TextBlockGoogleStyleFormatting]
Audit done.
```

looks like working as expected.
